packages/i18n/src/index.tsx

```diff
@@ -157,10 +158,98 @@ export const createI18nContext = (
      */
     dict: (lang: string) => deepReadObject(dict, lang)
   };
-  return [translate, actions];
+  return [translate, actions as any];
 };
 
 type I18nContextInterface = ReturnType<typeof createI18nContext>;
 
 export const I18nContext = createContext<I18nContextInterface>({} as I18nContextInterface);
 export const useI18n = () => useContext(I18nContext);
+
+// -------------------------- Chained I18n -------------------
+type I18nFormatArgs = Record<string, string | number>;
+
+export type I18nPath<T> = {
+  [K in keyof T]: T[K] extends Record<string, unknown>
+    ? I18nPath<T[K]>
+    : T[K] extends (...args: any) => string
+    ? (...args: Parameters<T[K]>) => string
+    : T[K] extends string
+    ? (args?: I18nFormatArgs) => string
+    : never;
+};
+
+const buildI18nChain = <T,>(obj: T): I18nPath<T> => {
+  const keys = Object.keys(obj as any) as (keyof T)[];
+  const paths = keys.reduce((acc, key) => {
+    const value = obj[key];
+    if (typeof value === "object") {
+      return {
+        ...acc,
+        [key]: buildI18nChain(value)
+      } as any;
+    } else if (typeof value === "function") {
+      return {
+        ...acc,
+        [key]: obj[key]
+      };
+    }
+    if (typeof value === "string") {
+      return {
+        ...acc,
+        [key]: (args: I18nFormatArgs) => {
+          return value.replace(/{{(.*?)}}/g, (_, key) => deepReadObject(args, key, ""));
+        }
+      };
+    } else {
+      throw new TypeError("Unsupported data format on the keys");
+    }
+  }, {} as Partial<I18nPath<T>>);
+
+  return paths as I18nPath<T>;
+};
+
+export const makeChainedI18nContext = <T extends object, K extends keyof T>(props: {
+  dictionaries: T;
+  locale: keyof T;
+  setContext?: boolean;
+}) => {
+  const [store, setStore] = createStore<{ dictionaries: T; locale: keyof T }>(props);
+
+  const utils = {
+    locale(locale?: K): K {
+      if (locale) {
+        setStore("locale", locale);
+        return locale;
+      }
+      return store.locale as K;
+    },
+    getDictionary(locale?: K): T[K] {
+      if (locale) return store.dictionaries[locale];
+      return store.dictionaries[store.locale as K];
+    }
+  };
+
+  const chainedI18n = createMemo(() => buildI18nChain(store.dictionaries[store.locale]));
+
+  const translate = () => {
+    return chainedI18n();
+  };
+
+  const context = createContext<[typeof translate, typeof utils] | null>(
+    props.setContext ? [translate, utils] : null
+  );
+
+  return {
+    I18nProvider: (props: { children: JSXElement }) => (
+      <context.Provider value={[translate, utils]}>{props.children}</context.Provider>
+    ),
+    useI18n: () => {
+      const i18nContext = useContext(context);
+      if (!i18nContext) {
+        throw new TypeError("useI18n must be used within a I18nProvider");
```

Comment:
why a `TypeError` instead of an `Error`?
Also the error message could be committed in the production build:
```suggestion
        throw new Error(process.env.DEV ? "useI18n must be used within a I18nProvider" : "");
```

Also, shouldn't the user choose how to handle the "context not found case"? Similarly to [this](https://github.com/solidjs-community/solid-primitives/blob/0620567359ec8bebdf3c99c9254960d4590c13e9/packages/i18n/src/index.tsx#L167). imo the primitives should do as little as possible, any additional functionality that could be extracted - should be extracted.
This style of handling missing context is very common so I would prefer to add a helper for it to [@solid-primitives/context](https://github.com/solidjs-community/solid-primitives/tree/main/packages/context#readme) and let the user compose it.

Reply:
Removed this error handle and updated my example to show wrapping it yourself if you want to handle it easier.